core/src/test/scala/integration/kafka/api/RackAwareReplicaAssignmentRackIdMapperTest.scala

```diff
@@ -0,0 +1,122 @@
+import org.junit.jupiter.api.{AfterEach, Assertions, Test}
+
+import java.util.Properties
+import scala.collection.JavaConverters._
+
+class RackAwareReplicaAssignmentRackIdMapperTest extends IntegrationTestHarness {
+  val rackIdMapperClassName = classOf[IgnorePrefixRackIdMapper].getCanonicalName
+  val rackIdMapper = CoreUtils.createObject[RackAwareReplicaAssignmentRackIdMapper](rackIdMapperClassName)
+  val brokerRacks = List(
+    "A::rack1",
+    "B::rack1",
+    "B::rack2",
+    "C::rack3",
+  )
+
+  var client: Admin = _
+
+  override val brokerCount = brokerRacks.size
+
+  @AfterEach
+  override def tearDown(): Unit = {
+    if (client != null) {
+      Utils.closeQuietly(client, "AdminClient")
+      client = null
+    }
+    super.tearDown()
+  }
+
+  override def modifyConfigs(configs: Seq[Properties]): Unit = {
+    super.modifyConfigs(configs)
+    // Specify customized rack.id
+    configs.zip(brokerRacks).foreach { case (cfgForEachBroker, rackId) =>
+      cfgForEachBroker.put(KafkaConfig.RackProp, rackId)
+      cfgForEachBroker.put(KafkaConfig.LiRackIdMapperClassNameForRackAwareReplicaAssignmentProp, rackIdMapperClassName)
+    }
+  }
+
+  def createConfig: java.util.Map[String, Object] = {
+    val config = new java.util.HashMap[String, Object]
+    config.put(AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG, brokerList)
+    config.put(AdminClientConfig.REQUEST_TIMEOUT_MS_CONFIG, "20000")
+    val securityProps: java.util.Map[Object, Object] =
+      adminClientSecurityConfigs(securityProtocol, trustStoreFile, clientSaslProperties)
+    securityProps.forEach { (key, value) => config.put(key.asInstanceOf[String], value) }
+    config
+  }
+
+  def waitForTopics(client: Admin, expectedPresent: Seq[String], expectedMissing: Seq[String]): Unit = {
+    waitUntilTrue(() => {
+      val topics = client.listTopics.names.get()
+      expectedPresent.forall(topicName => topics.contains(topicName)) &&
+        expectedMissing.forall(topicName => !topics.contains(topicName))
+    }, "timed out waiting for topics")
+  }
+
+
+  /**
+   * The rack is prefixed with some extra info, and the mapper should "ignore" the prefix.
+   * This is to simulate the case that we want to a) encode datacenter cages and b) group racks across cages
+   *
+   * If the mapper is not honored, with 8 partitions & RF=3, the underlying logic would place
+   *    8 (partitions) * 3 (RF) / 4 (brokers) = 6 replicas
+   * on each broker, because all of them are viewed as of different rack.
+   * The "mapped rack" rack1 has 2 brokers (0 & 1), which would hold 12 replicas under such situation,
+   * and by the pigeon hole theorem, there will be at least 1 partition having multiple replicas on the "mapped rack" rack1.
+   *
+   * On the other hand, if the mapper is honored, the broker will treat broker 0 & 1 as they're on the same rack
+   * and avoid placing the same partition on them.
+   */
+  @Test
+  def testCreateTopicWithExtraRackIdMapper(): Unit = {
```

Comment:
I think you may also need a test to verify that if all rack ids are different, the replicas should be placed evenly on the racks, and potentially also check all replicas on same rack does not belong to same partition.

Reply:
> I think you may also need a test to verify that if all rack ids are different, the replicas should be placed evenly on the racks
This one is covered in [existing tests](https://github.com/linkedin/kafka/blob/e227e027a592ad159d39b4ccc9bde1bbb02f4e17/core/src/test/scala/unit/kafka/admin/AdminRackAwareTest.scala).

>  check all replicas on same rack does not belong to same partition.
This one is covered here.  The toSet deduplication is to check if there are same partitions.  If there are, it will be deduplicated, and the size would be different.

Reply:
Ok should good as long as there are existing tests